File: android/app/src/main/java/com/claritygolf/caddy/wearables/garmin/GarminTransport.java
package com.claritygolf.caddy.wearables.garmin;

import android.content.Context;
import android.util.Log;

import com.claritygolf.caddy.BuildConfig;
import com.garmin.android.connectiq.ConnectIQ;
import com.garmin.android.connectiq.IQApp;
import com.garmin.android.connectiq.IQDevice;
import com.garmin.android.connectiq.exception.InvalidStateException;
import com.garmin.android.connectiq.exception.ServiceUnavailableException;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Talks to a Garmin wearable through the Garmin Connect Mobile app, using the
 * Connect IQ Mobile SDK (com.garmin.connectiq:ciq-companion-app-sdk, declared
 * in android/app/build.gradle). Called from
 * {@link com.claritygolf.caddy.NativeRoundBridge}, Android's Capacitor plugin.
 *
 * <p><b>Written against the real SDK as of 2026-09-20</b>, verified by
 * javap-ing the 2.4.0 AAR rather than from documentation. Until then every
 * call in here was commented out and written from inference; those inferences
 * were wrong in ways worth recording:
 *
 * <ul>
 *   <li>{@code IQApplicationEventListener.onMessageReceived} delivers a
 *       {@code List<Object>}, NOT the {@code Map} the old code tested for with
 *       {@code instanceof}. One watch message arrives as a list holding the
 *       Monkey C Dictionary. That shape would have compiled, run, and silently
 *       dropped every inbound command forever.</li>
 *   <li>Device status is {@code IQDevice.IQDeviceStatus} with four cases
 *       including {@code NOT_PAIRED} and {@code UNKNOWN}, not the two the old
 *       code collapsed to.</li>
 *   <li>Nearly every SDK call throws {@code InvalidStateException} (used
 *       before the SDK is ready) or {@code ServiceUnavailableException}
 *       (Garmin Connect Mobile missing, stopped or too old). Both are checked,
 *       so the old shape would not have compiled once uncommented.</li>
 * </ul>
 *
 * <p>Whether the app is installed on the watch is a real answer here, from
 * {@code getApplicationInfo}, rather than the "device is connected" proxy the
 * stub used. The two differ exactly when it matters: a connected watch with no
 * Clarity Caddy on it.
 *
 * <p>Mirrors GarminTransport.swift's responsibilities and talks directly to
 * whatever calls it, with no WearableCoordinator-style indirection: Android
 * has exactly one wearable target, so that abstraction (built for iOS to
 * arbitrate Apple Watch vs Garmin) has nothing to arbitrate here.
 *
 * <p>Does NOT implement a bytes-over-the-wire map asset path, for the same
 * reason GarminTransport.swift does not: Garmin pulls hole imagery by URL (see
 * garmin/source/Maps/GarminMapDownloader.mc) rather than receiving pushed
 * bytes, so publishMapManifest is the only map-related method here. The
 * manifest it forwards carries an absolute {@code url} per hole as of
 * 2026-09-19 (app/js/watch-map-delivery.js).
 *
 * <p>Package visibility needs nothing in our own manifest: the AAR declares
 * {@code <queries><package android:name="com.garmin.android.apps.connectmobile"/>}
 * itself, and manifest merging folds it in.
 */
public final class GarminTransport {

    /** Mirrors WearableTransportDelegate's shape from the iOS refactor, so a
     *  future Android NativeRoundBridge relays events to JavaScript the same
     *  way NativeRoundBridge.swift does via WearableCoordinatorDelegate. */
    public interface Listener {
        void onCommandReceived(Map<String, Object> command);
        void onMapInventoryReceived(Map<String, Object> inventory);
        void onPlayerInventoryReceived(Map<String, Object> inventory);
        void onStateChanged();
    }

    public static final class State {
        public final boolean supported;
        public final boolean activated;
        public final boolean paired;
        public final boolean appInstalled;
        public final boolean reachable;

        State(boolean supported, boolean activated, boolean paired, boolean appInstalled, boolean reachable) {
            this.supported = supported;
            this.activated = activated;
            this.paired = paired;
            this.appInstalled = appInstalled;
            this.reachable = reachable;
        }
    }

    private final Context context;
    private final GarminDeviceStore deviceStore;
    // The Connect IQ app identifier — must always equal garmin/manifest.xml's
    // <iq:application id="...">. See NativeRoundBridge.CONNECT_IQ_APP_ID for
    // why Android holds the undashed spelling and iOS the dashed one.
    private final String connectIqAppId;

    /* Garmin is a paid feature, and this is the gate that enforces it:
       send() refuses while it is false, so a membership that lapses stops the
       watch receiving rather than merely greying out a settings row.

       Defaults to FALSE and is only raised by JavaScript
       (NativeRoundBridge.setGarminEnabled, driven by
       ClarityPayments.hasActiveAccess). Failing closed is deliberate: if the
       payments module never loads we would rather one paying player reports a
       dead Garmin than every non-paying player quietly gets the feature.
       Volatile because setEntitled is called from the Capacitor bridge thread
       and read on whichever thread happens to be publishing. */
    private volatile boolean entitled = false;

    private Listener listener;

    private static final String TAG = "GarminTransport";

    private ConnectIQ connectIQ;
    private IQApp app;
    /** The device we are registered against. Held so we can unregister before
     *  registering a different one — the SDK keeps per-device listeners and
     *  would otherwise leak the old one and keep delivering its events. */
    private volatile IQDevice registeredDevice;
    /* Written from SDK callbacks (main thread), read from state() on the
       Capacitor bridge thread — same reasoning as `entitled` above. */
    private volatile boolean sdkReady;
    /** Real answer from getApplicationInfo, not "the device is connected".
     *  The two differ on a connected watch with no Clarity Caddy installed,
     *  which is exactly the case the pairing UI has to explain. */
    private volatile boolean appInstalledOnDevice;

    public GarminTransport(Context context, GarminDeviceStore deviceStore, String connectIqAppId) {
        this.context = context.getApplicationContext();
        this.deviceStore = deviceStore;
        this.connectIqAppId = connectIqAppId;
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    /** Brings the SDK up and, once it is ready, binds to whichever device the
     *  store currently holds. Safe to call repeatedly: initialize() is done
     *  once, and a later call just re-binds (which is what selectDevice wants
     *  after the player picks a different watch).
     *
     *  <p>autoUI = true lets the SDK put up Garmin's own dialogs when Garmin
     *  Connect Mobile is missing or too old. That is the right call here —
     *  those are the two failure modes a player can actually fix, and
     *  Garmin's wording for them is better than anything invented. */
    public void activate() {
        if (connectIQ == null) {
            connectIQ = ConnectIQ.getInstance(context, connectType());
            app = new IQApp(connectIqAppId);
        }
        if (sdkReady) { bindSelectedDevice(); return; }
        connectIQ.initialize(context, true, new ConnectIQ.ConnectIQListener() {
            @Override
            public void onSdkReady() {
                sdkReady = true;
                bindSelectedDevice();
                notifyStateChanged();
            }

            @Override
            public void onInitializeError(ConnectIQ.IQSdkErrorStatus status) {
                /* GCM_NOT_INSTALLED, GCM_UPGRADE_NEEDED or SERVICE_ERROR. The
                   first two are already on screen via autoUI; all three leave
                   us not ready, which state() reports honestly rather than
                   pretending the transport is live. */
                sdkReady = false;
                Log.w(TAG, "Connect IQ SDK did not initialise: " + status);
                notifyStateChanged();
            }

            @Override
            public void onSdkShutDown() {
                sdkReady = false;
                registeredDevice = null;
                appInstalledOnDevice = false;
                notifyStateChanged();
            }
        });
    }

    /** WIRELESS goes through Garmin Connect Mobile to a real watch. TETHERED
     *  is the SDK's simulator hook: it opens a socket on port 7381 and expects
     *  the desktop Connect IQ simulator on the other end of an adb forward, so
     *  the phone app and the watch app can be exercised together with no
     *  hardware. In that mode the SDK reports exactly one device, "Simulator",
     *  and getApplicationInfo answers for whatever the simulator is running.
     *
     *  <p>Chosen at build time (android/app/build.gradle, debug builds only)
     *  rather than at runtime so the choice cannot leak into a release. */
    private static ConnectIQ.IQConnectType connectType() {
        if (BuildConfig.GARMIN_TETHERED) {
            Log.i(TAG, "Connect IQ in TETHERED mode: expecting the simulator via `adb forward tcp:7381 tcp:7381`");
            return ConnectIQ.IQConnectType.TETHERED;
        }
        return ConnectIQ.IQConnectType.WIRELESS;
    }

    /** Releases the SDK. Paired with activate(); the selected device survives
     *  in the store, so coming back does not mean re-pairing. */
    public void deactivate() {
        if (connectIQ == null || !sdkReady) { return; }
        try {
            unregisterCurrentDevice();
            connectIQ.shutdown(context);
        } catch (InvalidStateException error) {
            /* Already down. Nothing to undo. */
        }
        sdkReady = false;
        registeredDevice = null;
        appInstalledOnDevice = false;
    }

    private void bindSelectedDevice() {
        GarminDeviceStore.SelectedDevice selected = deviceStore.getSelectedDevice();
        if (selected == null) { unregisterCurrentDevice(); return; }
        IQDevice device = toIQDevice(selected);
        if (device == null) { return; }
        if (registeredDevice != null && registeredDevice.getDeviceIdentifier() == device.getDeviceIdentifier()) { return; }
        unregisterCurrentDevice();
        try {
            connectIQ.registerForDeviceEvents(device, new ConnectIQ.IQDeviceEventListener() {
                @Override
                public void onDeviceStatusChanged(IQDevice changed, IQDevice.IQDeviceStatus status) {
                    deviceStore.recordConnectionState(connectionStateOf(status));
                    /* Whether the watch app is there can only be asked of a
                       connected device, so this is the moment to ask. */
                    if (status == IQDevice.IQDeviceStatus.CONNECTED) { refreshAppInstalled(changed); }
                    else { appInstalledOnDevice = false; }
                    notifyStateChanged();
                }
            });
            connectIQ.registerForAppEvents(device, app, new ConnectIQ.IQApplicationEventListener() {
                @Override
                public void onMessageReceived(IQDevice from, IQApp fromApp, List<Object> messages, ConnectIQ.IQMessageStatus status) {
                    if (status != ConnectIQ.IQMessageStatus.SUCCESS || messages == null) { return; }
                    /* A LIST, not a Map — one watch send arrives as a list
                       holding the Monkey C Dictionary. Testing the list itself
                       with `instanceof Map` (as the pre-SDK stub did) is false
                       every time and drops the message without a trace. */
                    for (Object message : messages) {
                        if (message instanceof Map) {
                            @SuppressWarnings("unchecked")
                            Map<String, Object> dictionary = (Map<String, Object>) message;
                            handleIncoming(dictionary);
                        }
                    }
                }
            });
            registeredDevice = device;
            deviceStore.recordConnectionState(connectionStateOf(currentStatusOf(device)));
            if (currentStatusOf(device) == IQDevice.IQDeviceStatus.CONNECTED) { refreshAppInstalled(device); }
        } catch (InvalidStateException | ServiceUnavailableException error) {
            Log.w(TAG, "could not bind Garmin device", error);
            registeredDevice = null;
        }
    }

    private void unregisterCurrentDevice() {
        if (connectIQ == null || registeredDevice == null) { return; }
        try { connectIQ.unregisterForEvents(registeredDevice); }
        catch (InvalidStateException error) { /* SDK already down */ }
        registeredDevice = null;
        appInstalledOnDevice = false;
    }

    private IQDevice.IQDeviceStatus currentStatusOf(IQDevice device) {
        try { return connectIQ.getDeviceStatus(device); }
        catch (InvalidStateException | ServiceUnavailableException error) { return IQDevice.IQDeviceStatus.UNKNOWN; }
    }

    private void refreshAppInstalled(IQDevice device) {
        try {
            connectIQ.getApplicationInfo(connectIqAppId, device, new ConnectIQ.IQApplicationInfoListener() {
                @Override
                public void onApplicationInfoReceived(IQApp installed) {
                    appInstalledOnDevice = true;
                    notifyStateChanged();
                }

                @Override
                public void onApplicationNotInstalled(String applicationId) {
                    appInstalledOnDevice = false;
                    notifyStateChanged();
                }
            });
        } catch (InvalidStateException | ServiceUnavailableException error) {
            appInstalledOnDevice = false;
        }
    }

    /** GarminDeviceStore keeps the id as a String because that is what crosses
     *  the JavaScript bridge; the SDK wants the long it really is. A value
     *  that is not a long cannot name a device, so it yields null rather than
     *  a device that will never match. */
    private IQDevice toIQDevice(GarminDeviceStore.SelectedDevice selected) {
        try {
            return new IQDevice(Long.parseLong(selected.deviceId), selected.deviceName);
        } catch (NumberFormatException error) {
            Log.w(TAG, "stored Garmin device id is not a device identifier: " + selected.deviceId);
            return null;
        }
    }

    private static GarminDeviceStore.ConnectionState connectionStateOf(IQDevice.IQDeviceStatus status) {
        if (status == IQDevice.IQDeviceStatus.CONNECTED) { return GarminDeviceStore.ConnectionState.CONNECTED; }
        if (status == IQDevice.IQDeviceStatus.NOT_PAIRED) { return GarminDeviceStore.ConnectionState.UNAVAILABLE; }
        if (status == IQDevice.IQDeviceStatus.NOT_CONNECTED) { return GarminDeviceStore.ConnectionState.NOT_CONNECTED; }
        return GarminDeviceStore.ConnectionState.UNKNOWN;
    }

    private void notifyStateChanged() {
        if (listener != null) { listener.onStateChanged(); }
    }

    public void publishScene(Map<String, Object> scene, Callback<Boolean> completion) {
        send(mapOf("scene", scene), completion);
    }

    /** See this class's header: {@code manifest} must carry a Garmin-specific
     *  `url` per hole for GarminMapDownloader to have anything to fetch.
     *  That attachment is not implemented here. */
    public void publishMapManifest(Map<String, Object> manifest, Callback<Boolean> completion) {
        send(mapOf("watchMapManifest", manifest), completion);
    }

    public void publishPlayer(Map<String, Object> player, Callback<Boolean> completion) {
        send(mapOf("watchPlayer", player), completion);
    }

    public void acknowledge(Map<String, Object> acknowledgement, Runnable completion) {
        send(mapOf("acknowledgement", acknowledgement), sent -> completion.run());
    }

    public State state() {
        GarminDeviceStore.SelectedDevice selected = deviceStore.getSelectedDevice();
        boolean connected = deviceStore.getLastKnownConnectionState() == GarminDeviceStore.ConnectionState.CONNECTED;
        return new State(
            /* supported: the SDK actually came up. False means Garmin Connect
               Mobile is missing, too old, or its service failed — all real,
               all worth showing differently from "no watch chosen". */
            sdkReady,
            /* activated: bound to a device and listening. */
            registeredDevice != null,
            /* paired: the player has chosen a watch. Survives disconnection
               and a membership lapse. */
            selected != null,
            /* appInstalled: the real answer from getApplicationInfo, not a
               proxy — a connected watch without Clarity Caddy reports false. */
            appInstalledOnDevice,
            connected
        );
    }

    public void setEntitled(boolean value) {
        if (entitled == value) { return; }
        entitled = value;
        /* A lapse does not clear the chosen device: the pairing survives and
           starts working again the moment access returns. Re-pairing after
           every billing hiccup would be its own bug. */
        if (listener != null) { listener.onStateChanged(); }
    }

    // ------------------------------------------- device selection (Settings)

    /** What the Settings > Garmin Watch page lists: the devices, plus whether
     *  the SDK is actually linked — "no devices" and "we cannot look" are
     *  different answers and the page words them differently. */
    public Map<String, Object> availableDevices() {
        HashMap<String, Object> out = new HashMap<>();
        ArrayList<Map<String, Object>> devices = new ArrayList<>();
        out.put("devices", devices);
        out.put("sdkLinked", true);
        if (connectIQ == null || !sdkReady) {
            /* Distinct from "looked and found none", and the settings page
               words it differently. activate() is called here so the common
               case — player opens the page before anything else has woken the
               SDK — resolves itself on their second tap rather than needing an
               app restart. */
            activate();
            out.put("reason", "Garmin Connect is not ready yet. Make sure the Garmin Connect app is installed and signed in, then try again.");
            return out;
        }
        try {
            /* Known, not connected: a watch that is paired in Garmin Connect
               but currently out of range is still the watch the player wants
               to choose. Its live status rides along so the page can say so. */
            for (IQDevice device : connectIQ.getKnownDevices()) {
                HashMap<String, Object> entry = new HashMap<>();
                entry.put("deviceId", String.valueOf(device.getDeviceIdentifier()));
                entry.put("deviceName", device.getFriendlyName());
                entry.put("model", partNumberOf(device));
                entry.put("connected", currentStatusOf(device) == IQDevice.IQDeviceStatus.CONNECTED);
                devices.add(entry);
            }
        } catch (InvalidStateException | ServiceUnavailableException error) {
            Log.w(TAG, "could not list Garmin devices", error);
            out.put("reason", "Could not reach Garmin Connect to list your watches.");
        }
        return out;
    }

    /** The device's part number, which is as close to a model name as the SDK
     *  offers. Best-effort: it needs a live service, and a missing model is
     *  cosmetic on the pairing row. */
    private String partNumberOf(IQDevice device) {
        try {
            String partNumber = connectIQ.getDevicePartNumber(device);
            return partNumber == null ? "" : partNumber;
        } catch (InvalidStateException | ServiceUnavailableException | IllegalArgumentException error) {
            return "";
        }
    }

    public void selectDevice(String deviceId, String deviceName, String model) {
        deviceStore.select(deviceId, deviceName, model);
        /* Re-binds listeners onto the newly chosen device; activate() is
           idempotent and unregisters the previous one first. */
        activate();
        notifyStateChanged();
    }

    public void clearSelectedDevice() {
        unregisterCurrentDevice();
        deviceStore.clearSelection();
        notifyStateChanged();
    }

    /** The richer state the settings page needs, over and above the five
     *  booleans {@link #state()} reports. */
    public Map<String, Object> garminState() {
        State current = state();
        java.util.HashMap<String, Object> out = new java.util.HashMap<>();
        out.put("supported", current.supported);
        out.put("activated", current.activated);
        out.put("paired", current.paired);
        out.put("appInstalled", current.appInstalled);
        out.put("reachable", current.reachable);
        out.put("entitled", entitled);
        // The SDK is bundled and this build talked to it: only false when
        // initialize() failed or has not run. A stub-era hard-coded false
        // survived here until 2026-09-20 and contradicted garminDevices().
        out.put("sdkLinked", sdkReady);
        out.put("connectionState", String.valueOf(deviceStore.getLastKnownConnectionState()));
        GarminDeviceStore.SelectedDevice selected = deviceStore.getSelectedDevice();
        if (selected != null) {
            java.util.HashMap<String, Object> device = new java.util.HashMap<>();
            device.put("deviceId", selected.deviceId);
            device.put("deviceName", selected.deviceName);
            device.put("model", selected.model);
            out.put("selectedDevice", device);
        }
        return out;
    }

    private void send(Map<String, Object> message, Callback<Boolean> completion) {
        // The paid gate, enforced below the web layer: no entitlement,
        // nothing leaves the phone.
        if (!entitled) { completion.onResult(false); return; }
        if (deviceStore.getSelectedDevice() == null) { completion.onResult(false); return; }
        if (connectIQ == null || !sdkReady || registeredDevice == null) { completion.onResult(false); return; }
        try {
            connectIQ.sendMessage(registeredDevice, app, message, new ConnectIQ.IQSendMessageListener() {
                @Override
                public void onMessageStatus(IQDevice device, IQApp sentApp, ConnectIQ.IQMessageStatus status) {
                    /* Reported, never inferred: only the SDK's own SUCCESS
                       counts as sent (Garmin Phase 1 plan step 8). The other
                       seven cases are all genuine failures, and
                       FAILURE_MESSAGE_TOO_LARGE in particular is one the
                       caller must see rather than have smoothed over — a
                       Scene that outgrew the link fails every time, not
                       intermittently. */
                    if (status != ConnectIQ.IQMessageStatus.SUCCESS) {
                        Log.w(TAG, "Garmin send failed: " + status);
                    }
                    completion.onResult(status == ConnectIQ.IQMessageStatus.SUCCESS);
                }
            });
        } catch (InvalidStateException | ServiceUnavailableException error) {
            Log.w(TAG, "Garmin send could not be attempted", error);
            completion.onResult(false);
        }
    }

    private void handleIncoming(Map<String, Object> message) {
        if (listener == null) { return; }
        Object command = message.get("command");
        if (command instanceof Map) { listener.onCommandReceived((Map<String, Object>) command); }
        Object mapInventory = message.get("watchMapHave");
        if (mapInventory instanceof Map) { listener.onMapInventoryReceived((Map<String, Object>) mapInventory); }
        Object playerInventory = message.get("watchPlayerHave");
        if (playerInventory instanceof Map) { listener.onPlayerInventoryReceived((Map<String, Object>) playerInventory); }
    }

    private static Map<String, Object> mapOf(String key, Object value) {
        java.util.HashMap<String, Object> m = new java.util.HashMap<>();
        m.put(key, value);
        return m;
    }

    public interface Callback<T> {
        void onResult(T value);
    }
}

File: ios/App/App/Wearables/Garmin/GarminTransport.swift
import Foundation
// import ConnectIQ  // Garmin Connect IQ Mobile SDK for iOS — not vendored in
                      // this repo yet. Add the .xcframework from Garmin's
                      // developer portal (developer.garmin.com/connect-iq),
                      // then uncomment this import and the commented
                      // conformances/calls below. This file IS in
                      // App.xcodeproj's Sources build phase and IS
                      // registered with WearableCoordinator
                      // (NativeRoundBridge.load()) — it compiles and runs
                      // today as a safe stub that never claims to reach a
                      // device (see `state()`/`send()` below), not because
                      // it's excluded from the build.

/*
 UNVERIFIED Connect IQ Mobile SDK calls — see below. WIRED INTO THE BUILD
 (Garmin Phase 1 plan step 4): registered alongside AppleWatchTransport in
 NativeRoundBridge.load(), and WearableCoordinator now fans out to both
 rather than only the first-registered transport — see
 WearableCoordinator.swift's header comment for why that had to change here
 rather than staying "deliberately left undefined."

 Everything that calls into `ConnectIQ`/`IQDevice`/`IQApp` below is written
 against this session's best understanding of the Connect IQ Mobile SDK for
 iOS's public API shape (ConnectIQ.sharedInstance() singleton; IQDevice for a
 paired device; IQApp scoping a message to one installed watch app;
 IQDeviceEventDelegate / IQAppMessageDelegate for callbacks). It has not been
 checked against the actual SDK headers — do that before trusting any single
 method or parameter name here. The ARCHITECTURE (what this class is
 responsible for, and the shape of WearableTransport it implements) is the
 part this session is confident about; the exact Garmin API calls are not.

 Mirrors AppleWatchTransport.swift's shape and responsibilities, adapted for
 Connect IQ:
   - initialise the Connect IQ SDK
   - discover/select a Garmin device (delegates the actual selection to
     GarminDeviceStore — this class only acts on whatever is already chosen)
   - register for device + app events
   - send/receive messages
   - expose state

 Does NOT own Caddy golf state — it is transport, exactly like
 AppleWatchTransport. Does NOT implement WearableFileAssetTransport: Garmin
 pulls map imagery by URL (see garmin/GarminMapDownloader.mc's header
 comment for why), so there is no bytes-over-the-wire asset path to
 implement on the phone side — publishMapManifest is enough, PROVIDED the
 manifest handed to this transport already carries a `url` per hole. That is
 not yet true of the manifest NativeRoundBridge.publishWatchMap forwards
 today (it is built once, generically, for whichever transports are
 registered) — wiring a Garmin-specific URL into that manifest is unresolved
 phone-side work, flagged here and in garmin/README.md, not silently assumed
 solved.
*/
final class GarminTransport: NSObject, WearableTransport {
    let platform: WearablePlatform = .garmin
    weak var delegate: WearableTransportDelegate?

    private let deviceStore: GarminDeviceStore
    private let queue = DispatchQueue(label: "com.claritygolf.caddy.garmin-transport")

    // The Connect IQ app identifier — the same UUID as garmin/manifest.xml's
    // <iq:application id="...">, supplied DASHED by NativeRoundBridge because
    // UUID(uuidString:) below rejects the undashed form the manifest uses (see
    // that constant's comment). Both must always agree: this is what scopes a
    // message to Caddy specifically among any other Connect IQ apps the paired
    // device might have.
    private let connectIQAppId: String

    private var latestScene: [String: Any]?

    init(deviceStore: GarminDeviceStore = GarminDeviceStore(), connectIQAppId: String) {
        self.deviceStore = deviceStore
        self.connectIQAppId = connectIQAppId
        super.init()
    }

    func activate() {
        // UNVERIFIED: ConnectIQ.sharedInstance().initialize(withUrlScheme:uiOverrideDelegate:)
        // or similar — the SDK needs a URL scheme registered in Info.plist
        // for the Connect Mobile app to hand control back to Caddy after
        // any device-pairing UI it presents. Also register for device
        // events on whichever device GarminDeviceStore currently holds, so
        // reachability/pairing changes reach `delegate` the same way
        // AppleWatchTransport's WCSessionDelegate callbacks do.
        //
        // if let selected = deviceStore.selectedDevice {
        //     let device = IQDevice(id: selected.deviceId, modelName: selected.model, friendlyName: selected.deviceName)
        //     ConnectIQ.sharedInstance().register(forDeviceEvents: device, delegate: self)
        //     let app = IQApp(uuid: UUID(uuidString: connectIQAppId), store: IQAppStore(), device: device)
        //     ConnectIQ.sharedInstance().register(forAppMessages: app, delegate: self)
        // }
    }

    // MARK: - WearableTransport

    func publishScene(_ scene: [String: Any], completion: @escaping (Bool) -> Void) {
        queue.async { [weak self] in
            guard let self else { return }
            self.latestScene = scene
            self.send(["scene": scene], completion: completion)
        }
    }

    func publishMapManifest(_ manifest: [String: Any], completion: @escaping (Bool) -> Void) {
        // See this file's header: `manifest` must carry a Garmin-specific
        // `url` per hole (garmin/GarminMapManifest.mc's `url` field) for
        // GarminMapDownloader to have anything to fetch. That attachment is
        // not implemented here — this method forwards whatever it is given.
        queue.async { [weak self] in
            guard let self else { return }
            self.send(["watchMapManifest": manifest], completion: completion)
        }
    }

    func publishPlayer(_ player: [String: Any], completion: @escaping (Bool) -> Void) {
        queue.async { [weak self] in
            guard let self else { return }
            self.send(["watchPlayer": player], completion: completion)
        }
    }

    func acknowledge(_ acknowledgement: [String: Any], completion: @escaping () -> Void) {
        queue.async { [weak self] in
            guard let self else { completion(); return }
            self.send(["acknowledgement": acknowledgement]) { _ in completion() }
        }
    }

    func state() -> WearableTransportState {
        let selected = deviceStore.selectedDevice
        let connected = deviceStore.lastKnownConnectionState == .connected
        return WearableTransportState(
            supported: true, // UNVERIFIED: should reflect ConnectIQ.sharedInstance() actually initialising
            activated: selected != nil,
            paired: selected != nil,
            appInstalled: connected, // best-effort proxy until device-status callbacks are wired
            reachable: connected
        )
    }

    // MARK: - Entitlement

    /* Garmin is a paid feature. This is the gate that actually enforces it:
       `send()` refuses while it is false, so a membership that lapses stops
       the watch receiving rather than merely greying out a settings row.

       It defaults to FALSE and is only ever raised by JavaScript
       (NativeRoundBridge.setGarminEnabled, driven by
       ClarityPayments.hasActiveAccess). Failing closed is deliberate: if the
       payments module never loads we would rather a paying player reports a
       dead Garmin than every non-paying player quietly gets the feature.
       Apple Watch is untouched by this — it has its own rules. */
    private var entitled = false

    func setEntitled(_ value: Bool) {
        queue.async { [weak self] in
            guard let self else { return }
            guard self.entitled != value else { return }
            self.entitled = value
            /* A lapse does not clear the chosen device. The player keeps their
               pairing and it starts working again the moment access returns —
               re-pairing after every billing hiccup would be its own bug. */
            self.delegate?.wearableTransportStateDidChange(self)
        }
    }

    // MARK: - Device selection (the Settings > Garmin Watch page)

    /* What the settings page lists. Returns the devices plus whether the SDK
       is actually linked, because "no devices" and "we cannot look" are
       different answers and the page says so in different words.

       UNVERIFIED / NOT YET POSSIBLE: the real implementation asks
       ConnectIQ.sharedInstance() for known devices, which on iOS means
       handing off to the Garmin Connect Mobile app and receiving the
       selection back through the registered URL scheme — there is no
       in-process device list to enumerate. Until the .xcframework is
       vendored this reports honestly that it cannot look. */
    func availableDevices() -> [String: Any] {
        // if ConnectIQ is linked:
        //   ConnectIQ.sharedInstance().showDeviceSelection()  // hands off to Garmin Connect
        //   and the chosen devices arrive via the URL-scheme callback, which
        //   should then call selectDevice(...) below.
        return [
            "devices": [[String: Any]](),
            "sdkLinked": false,
            "reason": "The Connect IQ Mobile SDK is not bundled in this build yet."
        ]
    }

    func selectDevice(id: String, name: String, model: String) {
        deviceStore.select(deviceId: id, deviceName: name, model: model)
        activate()
        delegate?.wearableTransportStateDidChange(self)
    }

    func clearSelectedDevice() {
        deviceStore.clearSelection()
        delegate?.wearableTransportStateDidChange(self)
    }

    /* The richer state the settings page needs, over and above the five
       booleans every transport reports through WearableTransportState. */
    func garminStateDictionary() -> [String: Any] {
        var out = state().asDictionary
        out["entitled"] = queue.sync { entitled }
        out["sdkLinked"] = false
        if let selected = deviceStore.selectedDevice {
            out["selectedDevice"] = [
                "deviceId": selected.deviceId,
                "deviceName": selected.deviceName,
                "model": selected.model
            ]
        }
        out["connectionState"] = deviceStore.lastKnownConnectionState.rawValue
        return out
    }

    // MARK: - Sending

    private func send(_ message: [String: Any], completion: @escaping (Bool) -> Void) {
        // The paid gate, enforced where it cannot be talked around from the
        // web layer: no entitlement, nothing leaves the phone.
        guard entitled else { completion(false); return }
        guard deviceStore.selectedDevice != nil else { completion(false); return }
        // UNVERIFIED: ConnectIQ.sharedInstance().sendMessage(_:toDevice:progress:completion:)
        // — the real send call. Until the SDK is linked this is a stub that
        // reports failure honestly rather than pretending to have sent
        // anything, matching the "native transport never infers success"
        // rule (Garmin Phase 1 plan step 8) — a stub must not claim
        // `published: true` it cannot back up.
        completion(false)
    }

    // MARK: - Receiving (wired once IQAppMessageDelegate is implemented)

    private func handleIncoming(_ message: [String: Any]) {
        if let command = message["command"] as? [String: Any] {
            delegate?.wearableTransport(self, didReceiveCommand: command)
        }
        if let inventory = message["watchMapHave"] as? [String: Any] {
            delegate?.wearableTransport(self, didReceiveMapInventory: inventory)
        }
        if let held = message["watchPlayerHave"] as? [String: Any] {
            delegate?.wearableTransport(self, didReceivePlayerInventory: held)
        }
    }
}

/*
 UNVERIFIED conformances, commented out until ConnectIQ types are linked:

 extension GarminTransport: IQDeviceEventDelegate {
     func deviceStatusChanged(_ device: IQDevice, status: IQDeviceStatus) {
         deviceStore.recordConnectionState(status == .connected ? .connected : .notConnected)
         delegate?.wearableTransportStateDidChange(self)
     }
 }

 extension GarminTransport: IQAppMessageDelegate {
     func receivedMessage(_ message: Any, from app: IQApp) {
         guard let dict = message as? [String: Any] else { return }
         handleIncoming(dict)
     }
 }
*/
